### PR TITLE
Implement FunctionHandle 

### PR DESCRIPTION
**Pull request by cemcayiroglu**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregationFunction.java [hunk 1]
```diff
@@ -91,7 +90,8 @@ protected AbstractMinMaxAggregationFunction(String name, boolean min)
     public InternalAggregationFunction specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionManager functionManager)
     {
         Type type = boundVariables.getTypeVariable("E");
-        MethodHandle compareMethodHandle = functionManager.getScalarFunctionImplementation(internalOperator(operatorType, BOOLEAN, ImmutableList.of(type, type))).getMethodHandle();
+        MethodHandle compareMethodHandle = functionManager.getScalarFunctionImplementation(
+                functionManager.resolveOperator(operatorType, TypeSignatureProvider.fromTypes(ImmutableList.of(type, type)))).getMethodHandle();
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregationFunction.java]: Static import `fromTypes`.

File: presto-main/src/main/java/com/facebook/presto/operator/index/FieldSetFilteringRecordSet.java [hunk 2]
```diff
@@ -55,7 +54,8 @@ public FieldSetFilteringRecordSet(FunctionManager functionManager, RecordSet del
             for (int field : fieldSet) {
                 fieldSetBuilder.add(new Field(
                         field,
-                        functionManager.getScalarFunctionImplementation(internalOperator(OperatorType.EQUAL, BooleanType.BOOLEAN, ImmutableList.of(columnTypes.get(field), columnTypes.get(field)))).getMethodHandle()));
+                        functionManager.getScalarFunctionImplementation(
+                                functionManager.resolveOperator(OperatorType.EQUAL, TypeSignatureProvider.fromTypes(ImmutableList.of(columnTypes.get(field), columnTypes.get(field))))).getMethodHandle()));
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/index/FieldSetFilteringRecordSet.java]: Static import `fromTypes`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/AbstractGreatestLeast.java [hunk 3]
```diff
@@ -104,7 +103,8 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
         Type type = boundVariables.getTypeVariable("E");
         checkArgument(type.isOrderable(), "Type must be orderable");
 
-        MethodHandle compareMethod = functionManager.getScalarFunctionImplementation(internalOperator(operatorType, BOOLEAN, ImmutableList.of(type, type))).getMethodHandle();
+        MethodHandle compareMethod = functionManager.getScalarFunctionImplementation(
+                functionManager.resolveOperator(operatorType, TypeSignatureProvider.fromTypes(type, type))).getMethodHandle();
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/AbstractGreatestLeast.java]: Same here

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapConstructor.java [hunk 4]
```diff
@@ -111,7 +110,8 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
         Type mapType = typeManager.getParameterizedType(MAP, ImmutableList.of(TypeSignatureParameter.of(keyType.getTypeSignature()), TypeSignatureParameter.of(valueType.getTypeSignature())));
         MethodHandle keyHashCode = functionManager.getScalarFunctionImplementation(functionManager.resolveOperator(OperatorType.HASH_CODE, fromTypes(keyType))).getMethodHandle();
         MethodHandle keyEqual = functionManager.getScalarFunctionImplementation(functionManager.resolveOperator(OperatorType.EQUAL, fromTypes(keyType, keyType))).getMethodHandle();
-        MethodHandle keyIndeterminate = functionManager.getScalarFunctionImplementation(internalOperator(INDETERMINATE.name(), BOOLEAN.getTypeSignature(), ImmutableList.of(keyType.getTypeSignature()))).getMethodHandle();
+        MethodHandle keyIndeterminate = functionManager.getScalarFunctionImplementation(
+                functionManager.resolveOperator(INDETERMINATE, TypeSignatureProvider.fromTypeSignatures((keyType.getTypeSignature())))).getMethodHandle();
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapConstructor.java]: static import

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RowIndeterminateOperator.java [hunk 5]
```diff
@@ -132,15 +132,14 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
...
-                ScalarFunctionImplementation function = functionManager.getScalarFunctionImplementation(signature);
+                FunctionHandle functionHandle = functionManager.resolveOperator(
+                        INDETERMINATE,
+                        TypeSignatureProvider.fromTypeSignatures(fieldTypes.get(i).getTypeSignature()));
+                ScalarFunctionImplementation function = functionManager.getScalarFunctionImplementation(functionHandle);
                 BytecodeExpression element = constantType(binder, fieldTypes.get(i)).getValue(value, constantInt(i));
 
                 ifNullField.ifFalse(new IfStatement("if the field is not null but indeterminate...")
-                        .condition(invokeFunction(scope, cachedInstanceBinder, signature.getName(), function, element))
+                        .condition(invokeFunction(scope, cachedInstanceBinder, functionHandle.getSignature().getName(), function, element))
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RowIndeterminateOperator.java]: You can probably use `mangleOperatorName(INDETERMINATE)` to get the name here. There's no need to resolve the operator here.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregationFunction.java [hunk 6]
```diff
@@ -91,7 +90,8 @@ protected AbstractMinMaxAggregationFunction(String name, boolean min)
     public InternalAggregationFunction specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionManager functionManager)
     {
         Type type = boundVariables.getTypeVariable("E");
-        MethodHandle compareMethodHandle = functionManager.getScalarFunctionImplementation(internalOperator(operatorType, BOOLEAN, ImmutableList.of(type, type))).getMethodHandle();
+        MethodHandle compareMethodHandle = functionManager.getScalarFunctionImplementation(
+                functionManager.resolveOperator(operatorType, fromTypes(ImmutableList.of(type, type)))).getMethodHandle();
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregationFunction.java]: You can just use `fromTypes(type, type)`

File: presto-main/src/main/java/com/facebook/presto/operator/index/FieldSetFilteringRecordSet.java [hunk 7]
```diff
@@ -55,7 +54,8 @@ public FieldSetFilteringRecordSet(FunctionManager functionManager, RecordSet del
             for (int field : fieldSet) {
                 fieldSetBuilder.add(new Field(
                         field,
-                        functionManager.getScalarFunctionImplementation(internalOperator(OperatorType.EQUAL, BooleanType.BOOLEAN, ImmutableList.of(columnTypes.get(field), columnTypes.get(field)))).getMethodHandle()));
+                        functionManager.getScalarFunctionImplementation(
+                                functionManager.resolveOperator(OperatorType.EQUAL, fromTypes(ImmutableList.of(columnTypes.get(field), columnTypes.get(field))))).getMethodHandle()));
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/index/FieldSetFilteringRecordSet.java]: Same here.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapElementAtFunction.java [hunk 8]
```diff
@@ -86,7 +85,8 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
         Type keyType = boundVariables.getTypeVariable("K");
         Type valueType = boundVariables.getTypeVariable("V");
 
-        MethodHandle keyEqualsMethod = functionManager.getScalarFunctionImplementation(internalOperator(OperatorType.EQUAL, BooleanType.BOOLEAN, ImmutableList.of(keyType, keyType))).getMethodHandle();
+        MethodHandle keyEqualsMethod = functionManager.getScalarFunctionImplementation(
+                functionManager.resolveOperator(OperatorType.EQUAL, TypeSignatureProvider.fromTypes(keyType, keyType))).getMethodHandle();
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapElementAtFunction.java]: static import `fromTypes`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapHashCodeOperator.java [hunk 9]
```diff
@@ -56,8 +55,10 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
         Type keyType = boundVariables.getTypeVariable("K");
         Type valueType = boundVariables.getTypeVariable("V");
 
-        MethodHandle keyHashCodeFunction = functionManager.getScalarFunctionImplementation(internalOperator(HASH_CODE, BIGINT, ImmutableList.of(keyType))).getMethodHandle();
-        MethodHandle valueHashCodeFunction = functionManager.getScalarFunctionImplementation(internalOperator(HASH_CODE, BIGINT, ImmutableList.of(valueType))).getMethodHandle();
+        MethodHandle keyHashCodeFunction = functionManager.getScalarFunctionImplementation(
+                functionManager.resolveOperator(HASH_CODE, TypeSignatureProvider.fromTypes(keyType))).getMethodHandle();
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapHashCodeOperator.java]: static import

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapHashCodeOperator.java [hunk 10]
```diff
@@ -56,8 +55,10 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
...
+        MethodHandle valueHashCodeFunction = functionManager.getScalarFunctionImplementation(
+                functionManager.resolveOperator(HASH_CODE, TypeSignatureProvider.fromTypes(valueType))).getMethodHandle();
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapHashCodeOperator.java]: static import

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RowIndeterminateOperator.java [hunk 11]
```diff
@@ -132,15 +133,14 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
...
                                 .gotoLabel(end));
 
-                Signature signature = internalOperator(
-                        INDETERMINATE.name(),
-                        BOOLEAN.getTypeSignature(),
-                        ImmutableList.of(fieldTypes.get(i).getTypeSignature()));
-                ScalarFunctionImplementation function = functionManager.getScalarFunctionImplementation(signature);
+                FunctionHandle functionHandle = functionManager.resolveOperator(
+                        INDETERMINATE,
+                        TypeSignatureProvider.fromTypeSignatures(fieldTypes.get(i).getTypeSignature()));
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RowIndeterminateOperator.java]: static import

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RowIndeterminateOperator.java [hunk 12]
```diff
@@ -132,15 +133,14 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
...
+                ScalarFunctionImplementation function = functionManager.getScalarFunctionImplementation(functionHandle);
                 BytecodeExpression element = constantType(binder, fieldTypes.get(i)).getValue(value, constantInt(i));
 
                 ifNullField.ifFalse(new IfStatement("if the field is not null but indeterminate...")
-                        .condition(invokeFunction(scope, cachedInstanceBinder, signature.getName(), function, element))
+                        .condition(invokeFunction(scope, cachedInstanceBinder, mangleOperatorName(INDETERMINATE), function, element))
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RowIndeterminateOperator.java]: Let's use `functionHandle.getSignature().getName()` for now. Using `mangleOperatorName` directly feels hacky.

**Comment by cemcayiroglu** [comment]
closing since is it is a duplicate.

**Pull request closed without merging**